omap-generator: Fix memory problem in the csi controller cache

Madhu-1 opened the pull request:
#  Watch for only metadata changes in PV
    
In the omap controller we are watching for the PV object and also we use controller runtime client to get the secrets. There are two problems with it.    
1. Watching for the entire PV object has drawbacks in the larger clusters where the entire PV object will be stored in the cache which inturn increases the memory consumption.
2. When we do get on the controller runtime client it also starts a informer for the object, As we are getting the secrets to talk to ceph. the informer is created and it caches in the secrets as well.

Because of these 2 problem. In a cluster where i have around 20K PVC and 10K secrets. The csi omap generator controller memory went to 4Gib.    
 To fix these problem. for first problem Cache only the PV metadata object by watching for only metadata changes. To fix the second problem, Disable the cache for both PV and the secrets in the client options.
 This increases the number of Get Calls to the API server but it avoids the List/Watch on the secret. and an additional get call will be made of a PV object.
    
With this change in the same cluster, the memory consumption went from 4Gib to 200Mb for the omap-generator controller.

# add predicate to avoid Reconcile
    
We can avoid reconcilation in multiple occasion.
1. The PV controller adds the csi driver name as the annotation in the PV object we can check for it and avoid reconcile if the driver is not matching.
2. When the PV is under deletion
3. When there is a delete event


## Memory consumption before

```bash
[🎩︎]mrajanna@memory-test $]oc adm top pod -nopenshift-storage openshift-storage.rbd.csi.ceph.com-ctrlplugin-988d85874-mcbhm --containers
POD                                                             NAME                 CPU(cores)   MEMORY(bytes)   
openshift-storage.rbd.csi.ceph.com-ctrlplugin-988d85874-mcbhm   csi-addons           0m           22Mi            
openshift-storage.rbd.csi.ceph.com-ctrlplugin-988d85874-mcbhm   csi-attacher         1m           168Mi           
openshift-storage.rbd.csi.ceph.com-ctrlplugin-988d85874-mcbhm   csi-omap-generator   180m         3465Mi          
openshift-storage.rbd.csi.ceph.com-ctrlplugin-988d85874-mcbhm   csi-provisioner      31m          637Mi           
openshift-storage.rbd.csi.ceph.com-ctrlplugin-988d85874-mcbhm   csi-rbdplugin        1m           97Mi            
openshift-storage.rbd.csi.ceph.com-ctrlplugin-988d85874-mcbhm   csi-resizer          2m           285Mi           
openshift-storage.rbd.csi.ceph.com-ctrlplugin-988d85874-mcbhm   csi-snapshotter      1m           17Mi            
openshift-storage.rbd.csi.ceph.com-ctrlplugin-988d85874-mcbhm   log-rotator          0m           1Mi
```

## Memory consumption after

```bash
[🎩︎]mrajanna@memory-test $]oc adm top pod openshift-storage.rbd.csi.ceph.com-ctrlplugin-66f7db5bc8-xrclg --containers
POD                                                              NAME                 CPU(cores)   MEMORY(bytes)   
openshift-storage.rbd.csi.ceph.com-ctrlplugin-66f7db5bc8-xrclg   csi-addons           1m           18Mi            
openshift-storage.rbd.csi.ceph.com-ctrlplugin-66f7db5bc8-xrclg   csi-attacher         1m           12Mi            
openshift-storage.rbd.csi.ceph.com-ctrlplugin-66f7db5bc8-xrclg   csi-omap-generator   161m         211Mi           
openshift-storage.rbd.csi.ceph.com-ctrlplugin-66f7db5bc8-xrclg   csi-provisioner      1m           488Mi           
openshift-storage.rbd.csi.ceph.com-ctrlplugin-66f7db5bc8-xrclg   csi-rbdplugin        1m           24Mi            
openshift-storage.rbd.csi.ceph.com-ctrlplugin-66f7db5bc8-xrclg   csi-resizer          2m           327Mi           
openshift-storage.rbd.csi.ceph.com-ctrlplugin-66f7db5bc8-xrclg   csi-snapshotter      1m           16Mi            
openshift-storage.rbd.csi.ceph.com-ctrlplugin-66f7db5bc8-xrclg   log-rotator          1m           1Mi  
```

Madhu-1 commented:
/test ci/centos/mini-e2e-helm/k8s-1.31

nixpanic commented on internal/controller/persistentvolume/persistentvolume.go:
```diff
@@ -168,17 +170,54 @@ func newPVReconciler(mgr manager.Manager, config ctrl.Config) reconcile.Reconcil
+		return true
+	}
+
+	if !obj.GetDeletionTimestamp().IsZero() {
```
nit: move the `obj.GetDeletionTimestamp().IsZero()` check before the annotation get? That groups all annotation checks together.

nixpanic reviewed (commented):
The 1st commit seems to address two distinct things; it is cleaner to split that in two commits.

Nit: ordering of statements in `shouldReconcileBasedOnDriver()` is a little weird.

Madhu-1 committed:
csi: disable cache for PV and secret

When we do get on the controller runtime client
it also starts a informer for the object, For ex as
we are getting the secrets. the
informer is created and it caches in the secrets
as well.

In a cluster where i have
around 20K PVC and 10K secrets. The csi omap generator
controller memory went to 4Gib.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
csi: Watch for only metadata changes in PV

Wathching for the entire PV object has drawbacks
in the larger clusters where the entire PV object
will be stored in the cache which inturn increases
the memory consumption.

To fix these problem. for first problem Cache only the
PV metadata object by watching for only metadata changes.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
csi: add predicate to avoid Reconcile

We can avoid reconcilation in multiple
occasion.
1. The PV controller adds the csi driver
name as the annotation in the PV object
we can check for it and avoid reconcile
if the driver is not matching
2. When the PV is under deletion
3. When there is a delete event

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Rakshith-R commented on internal/controller/persistentvolume/persistentvolume.go:
```diff
@@ -168,15 +170,52 @@ func newPVReconciler(mgr manager.Manager, config ctrl.Config) reconcile.Reconcil
 }
 
-func add(mgr manager.Manager, r reconcile.Reconciler) error {
+func shouldReconcileBasedOnDriver(obj client.Object, driverName string) bool {
```
Please add a comment for this function and unit tests too,

iPraveenParihar commented on internal/controller/persistentvolume/persistentvolume.go:
```diff
@@ -169,26 +169,16 @@ func newPVReconciler(mgr manager.Manager, config ctrl.Config) reconcile.Reconcil
+	// Create a new controller builder
+	build := builder.ControllerManagedBy(mgr).Named(
+		"persistentvolume-controller").WithOptions(
+		controller.Options{MaxConcurrentReconciles: 1})
```
```suggestion
		controller.Options{MaxConcurrentReconciles: 1, Reconciler: r})
```
Reconciler not required here?

Madhu-1 replied:
its in complete below

Rakshith-R reviewed (commented):
@Madhu-1 Let's optimise VGRC cache and predicate too ?

Madhu-1 commented:
> @Madhu-1 Let's optimise VGRC cache and predicate too ?

Not part of this PR, I will address it in followup. it needs lot of testing as well :)

iPraveenParihar reviewed (approved):
LGTM, lets have a tracker for optimising VolumeGroupReplicationContent controller.

Rakshith-R reviewed (approved):
LGTM, Thanks

Madhu-1 commented:
@mergifyio queue
